dxr/plugins/rust/menu.py

```diff
@@ -34,106 +35,187 @@ def truncate_value(value, typ=""):
+    def menu_items(self):
+        qualname = self.data
+        return [{'html': "Find callers",
+                 'title': "Find functions that call this function",
+                 'href': search_url(self.tree, "+callers:%s" % quote(qualname)),
+                 'icon': 'method'},
+                {'html': "Find callees",  # TODO: Probably useless. Remove.
+                 'title': "Find functions that are called by this function",
+                 'href': search_url(self.tree, "+called-by:%s" % quote(qualname)),
+                 'icon': 'method'}]
+
+
+# TODO: DRY with clang plugin.
+class _JumpToTarget(MultiDatumMenuMaker):
+    """A MenuMaker that jumps straight to a specific line of a file
+
+    :classvar target_name: A description of the kind of thing I point to, like
+        "trait method" or "function definition"
+
+    """
+    def __init__(self, tree, path, row):
+        super(_JumpToTarget, self).__init__(tree)
+        self.path = path
+        self.row = row
+
+    def es(self):
+        return self.path, self.row
+
+    @classmethod
+    def from_decl(cls, tree, decl):
+        """Return an instance made from a declaration mapping.
+
+        If the incoming declaration doesn't warrant the creation of a menu,
+        return None.
+
+        """
+        path = decl['file_name']
+        if path:
+            return cls(tree, path, decl['file_line'])
+        else:
+            warn("Can't add jump to empty path.")  # Can this happen?
+
+    def menu_items(self):
+        yield {'html': 'Jump to %s' % self.target_name,
+               'title': "Jump to %s in '%s'" % (self.target_name,
+                                                os.path.basename(self.path)),
+               'href': url_for(BROWSE, tree=self.tree.name, path=self.path, _anchor=self.row),
+               'icon': 'jump'}
+
+
+class JumpToTraitMethodMenuMaker(_JumpToTarget, _RustPluginAttr):
+    target_name = 'trait method'
+
+
+class JumpToDefinitionMenuMaker(_JumpToTarget, _RustPluginAttr):
+    target_name = 'definition'
+
+
+class JumpToModuleDefinitionMenuMaker(_JumpToTarget, _RustPluginAttr):
+    target_name = 'module defintion'
+
+
+class JumpToModuleDeclarationMenuMaker(_JumpToTarget, _RustPluginAttr):
+    target_name = 'module declaration'
 
-def function_menu(tree, datum, tree_config):
-    menu = function_menu_generic(tree, datum, tree_config)
 
+class JumpToAliasDefinitionMenuMaker(_JumpToTarget, _RustPluginAttr):
+    target_name = 'alias defintion'
+
+
+class JumpToCrateMenuMaker(_JumpToTarget, _RustPluginAttr):
+    target_name = 'crate'
+
+
+class JumpToTypeDeclarationMenuMaker(_JumpToTarget, _RustPluginAttr):
+    target_name = 'type declaration'
+
+
+class JumpToVariableDeclarationMenuMaker(_JumpToTarget, _RustPluginAttr):
+    target_name = 'variable declaration'
+
+
+class JumpToFunctionDeclarationMenuMaker(_JumpToTarget, _RustPluginAttr):
+    target_name = 'function declaration'
```

Comment:
Why all these classes? Seems like it would be nicer to make target_name a regular field set in the ctor.